### PR TITLE
Improvements to the DynamicArray class

### PR DESCRIPTION
**Pull request by JulienIgnace-Unity**
### Purpose of this PR
This PR adds a few utility APIs to the DynamicArray class. The goal is to provide a serviceable replacement for List when allocations are not allowed. ICollection interface allocates all over the place, in foreach, Sort etc and in some critical code that can be called every frame, we can't afford that.

---
### Testing status
Automated tests have been added for all APIs.

---
### Comments to reviewers
Added @arttu-peltonen for a Core SRP perspective on this.
Also, the QuickSort was implemented as a separate static function to avoid the requirement of IComparable for ALL dynamic arrays. Not super happy with that so if anyone has a more elegant solution I'm all for it :)

### CODE REVIEW

File: com.unity.render-pipelines.core/Runtime/Common/DynamicArray.cs [hunk 1]
```diff
@@ -78,21 +212,88 @@ public int Add(in T value)
...
+
+                if (pivot >= 1)
+                    QuickSort(data, left, pivot);
+
+                if (pivot + 1 < right)
+                    QuickSort(data, pivot + 1, right);
+            }
+        }
+
+        public static void QuickSort<U>(DynamicArray<U> array) where U : IComparable<U>, new()
```
Comment (arttu-peltonen) [inline com.unity.render-pipelines.core/Runtime/Common/DynamicArray.cs:294]: You should be able to achieve better API by making this an [extension method](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/extension-methods). For that, it's needs to be declared in a separate static non-nested class, naming convention here would be `public static class DynamicArrayExtensions`. Signature something like this: `public static void QuickSort<T>(this DynamicArray<T> array) where T : IComparable<T>, new()`. That way you can keep the stricter type requirements for this function only, but calling it looks like just calling a normal method: `m_DynamicArray.QuickSort();`

**Review by arttu-peltonen — CHANGES_REQUESTED** [review]
Looks great! I recommend just making the QuickSort change to keep the API consistent and make the method more discoverable via autocomplete.

File: com.unity.render-pipelines.core/Runtime/Common/DynamicArray.cs (hunk 1 above)
Comment (JulienIgnace-Unity) [reply com.unity.render-pipelines.core/Runtime/Common/DynamicArray.cs:294]: I did not think of that. I knew it was a good idea to add you as a reviewer :)
Will try to implement that.

Comment (JulienIgnace-Unity) [reply com.unity.render-pipelines.core/Runtime/Common/DynamicArray.cs]: Updated the PR with the extension. Looks much cleaner now.

File: com.unity.render-pipelines.core/Runtime/Common/DynamicArray.cs [hunk 2]
```diff
@@ -70,6 +81,129 @@ public int Add(in T value)
...
 
+        /// <summary>
+        /// Adds the elements of the specified collection to the end of the DynamicArray.
+        /// </summary>
+        /// <param name="array">The array whose elements should be added to the end of the DynamicArray. The array itself cannot be null, but it can contain elements that are null, if type T is a reference type.</param>
+        public void AddRange(DynamicArray<T> array)
+        {
+            Reserve(size + array.size);
+            for (int i = 0; i < array.size; ++i)
+                Add(array[i]);
```
Comment (FrancescoC-unity) [inline com.unity.render-pipelines.core/Runtime/Common/DynamicArray.cs:92]: Quick performance note, given that we do the reserve before, there should be no need of doing the full Add that does range check,  unless I missed something.

**Review by FrancescoC-unity — APPROVED** [review]
LGTM except a small perf improvement that can be done

File: com.unity.render-pipelines.core/Runtime/Common/DynamicArray.cs (hunk 2 above)
Comment (JulienIgnace-Unity) [reply com.unity.render-pipelines.core/Runtime/Common/DynamicArray.cs]: Good point. Updated the PR.

**Review by sebastienlagarde — APPROVED** [review]

**Pull request merged**